framework/src/variables/MooseLinearVariableFV.C

```diff
@@ -0,0 +1,263 @@
+  : MooseVariableField<OutputType>(parameters),
+    _solution(this->_sys.currentSolution()),
+    _needs_cell_gradients(false)
+{
+}
+
+template <typename OutputType>
+bool
+MooseLinearVariableFV<OutputType>::isExtrapolatedBoundaryFace(
+    const FaceInfo & /*fi*/, const Elem * const /*elem*/, const Moose::StateArg & /*state*/) const
+{
+  /// This is not used by this variable at this point.
+  return false;
+}
+
+template <typename OutputType>
+Moose::VarFieldType
+MooseLinearVariableFV<OutputType>::fieldType() const
+{
+  if (std::is_same<OutputType, Real>::value)
+    return Moose::VarFieldType::VAR_FIELD_STANDARD;
+  else if (std::is_same<OutputType, RealVectorValue>::value)
+    return Moose::VarFieldType::VAR_FIELD_VECTOR;
+  else if (std::is_same<OutputType, RealEigenVector>::value)
+    return Moose::VarFieldType::VAR_FIELD_ARRAY;
+  else
+    mooseError("Unknown variable field type");
+}
+
+template <typename OutputType>
+bool
+MooseLinearVariableFV<OutputType>::isArray() const
+{
+  return std::is_same<OutputType, RealEigenVector>::value;
+}
+
+template <typename OutputType>
+bool
+MooseLinearVariableFV<OutputType>::isVector() const
+{
+  return std::is_same<OutputType, RealVectorValue>::value;
+}
+
+template <typename OutputType>
+Real
+MooseLinearVariableFV<OutputType>::getElemValue(const ElemInfo * const elem_info,
+                                                const StateArg & state) const
+{
+  mooseAssert(
+      this->hasBlocks(elem_info->subdomain_id()),
+      "The variable should be defined on the element's subdomain! This typically occurs when the "
+      "user wants to evaluate the elements right next to the boundary of two variables (block "
+      "boundary). The subdomain which is queried: " +
+          Moose::stringify(this->activeSubdomains()) + " the subdomain of the element " +
+          std::to_string(elem_info->subdomain_id()));
+
+  // It's not safe to use solutionState(0) because it returns the libMesh System solution member
+  // which is wrong during things like finite difference Jacobian evaluation, e.g. when PETSc
+  // perturbs the solution vector we feed these perturbations into the current_local_solution
+  // while the libMesh solution is frozen in the non-perturbed state
+  const auto & global_soln = (state.state == 0)
+                                 ? *this->_sys.currentSolution()
+                                 : this->_sys.solutionState(state.state, state.iteration_type);
+
+  return global_soln(elem_info->dofIndices()[this->_sys.number()][this->number()]);
+}
+
+template <typename OutputType>
+const VectorValue<Real> &
+MooseLinearVariableFV<OutputType>::gradSln(const ElemInfo * const elem_info) const
+{
+  if (_needs_cell_gradients)
+  {
+    _cell_gradient.zero();
+    for (const auto i : make_range(this->_mesh.dimension()))
+      _cell_gradient(i) =
+          (*_grad_cache[i])(elem_info->dofIndices()[this->_sys.number()][this->number()]);
+  }
+
+  return _cell_gradient;
```

Comment:
I will change this, I just thought I could save some reallocation here and remembered that we return references in the original FV variable due to potentially expensive AD coy operations.

Reply:
yes I realized that we have the same danger in the nonlinear FV variable class when not caching cell gradients. We probably get away with it because the default is to cache and we store the cached values in a global map. We should probably just return a value instead of a reference from `adGradSln` for safety there as well